packages/gatsby/src/commands/serve.js

```diff
@@ -40,6 +41,29 @@ const readMatchPaths = async program => {
   return JSON.parse(rawJSON)
 }
 
+/**
+ * Adds trailing slash to request.url, in order to deceive static middleware
+ * into getting a directory request, thus preventing redirect.
+ * This is an express middleware to be used before static middleware
+ * @param {*} req
+ * @param {*} res
+ * @param {*} next
+ */
+const directoryRedirectPrevention = (req, res, next) => {
+  // url is home page
+  if (req.url === `/`) return next()
+  // url has trailing slash, no need to intervene
+  if (req.url.slice(-1) === `/`) return next()
+  // url is a file request, no need for trailing slash
+  if (fileExtensionRegEx.test(req.url)) return next()
+
+  // turning /slug into /slug/
+  req.originalUrl += `/` // https://expressjs.com/en/api.html#req.originalUrl
```

Comment:
The first example in this link has:
```
// GET /search?q=something
console.dir(req.originalUrl)
// => '/search?q=something'
```

So potentially this line would be doing `originalUrl = '/search?q=something' + '/'`. That's probably not what you wanted?

Reply:
Indeed, url's with query args is a use case I haven't tackled yet. I'm not fully familiar with gatsby's handle of query args. Your input here is required.
Do we turn `/search?q=something` into `/search/?q=something` ?

Reply:
To be honest, I have no idea :)

Maybe @pieh knows

Reply:
We have this piece of code in production runtime that adjust url to use canonical page path for initial page load (so this will be connected with to `gatsby serve` behaviour in some way) - https://github.com/gatsbyjs/gatsby/blob/9d3a8f080ca0a9d58a92a978931b3438f4d64e0d/packages/gatsby/cache-dir/production-app.js#L99-L119

Just note that "canonical path" on its own doesn't specify if it has trailing slash or not - it just mean `path` that was used with `createPage` action call (Gatsby doesn't normalize those right now - at least for pages created in user code in `createPages` hook). Gatsby does however use trailing slashes for pages created from `src/pages` directory (so it does have opinion here).

I'm not sure if this make any sense or if this is answer any question here.

But for manipulating URLs I recommend using builtin Node's `url` module - particularly:
 - [url.parse](https://nodejs.org/api/url.html#url_url_parse_urlstring_parsequerystring_slashesdenotehost) - to created structured url object
 - then modify `pathname` part to add trailing slash
 - [url.format](https://nodejs.org/api/url.html#url_url_format_urlobject) - for converting object back to string